topology/probes/k8s/cluster.go

```diff
@@ -27,47 +27,79 @@ import (
 	"github.com/skydive-project/skydive/probe"
 	"github.com/skydive-project/skydive/topology"
 	"github.com/skydive-project/skydive/topology/graph"
-
-	"k8s.io/client-go/kubernetes"
 )
 
 // ClusterName is the name of the k8s cluster
 const ClusterName = "cluster"
 
 var clusterEventHandler = graph.NewEventHandler(100)
 
+var clusterNode *graph.Node
+
 type clusterCache struct {
 	*graph.EventHandler
 	graph *graph.Graph
 }
 
-func (c *clusterCache) Start() {
+func (c *clusterCache) addClusterNode() {
 	c.graph.Lock()
 	defer c.graph.Unlock()
 
 	m := NewMetadata(Manager, "cluster", nil, ClusterName)
-	node := c.graph.NewNode(graph.GenID(), m, "")
-	c.NotifyEvent(graph.NodeAdded, node)
+	clusterNode = c.graph.NewNode(graph.GenID(), m, "")
+	c.NotifyEvent(graph.NodeAdded, clusterNode)
 	logging.GetLogger().Debugf("Added cluster{Name: %s}", ClusterName)
 }
 
+func (c *clusterCache) Start() {
+}
+
 func (c *clusterCache) Stop() {
 }
 
-func newClusterProbe(clientset *kubernetes.Clientset, g *graph.Graph) Subprobe {
-	return &clusterCache{
+func newClusterProbe(clientset interface{}, g *graph.Graph) Subprobe {
+	c := &clusterCache{
 		EventHandler: clusterEventHandler,
 		graph:        g,
 	}
+	c.addClusterNode()
```

Comment:
This can be moved to the Start method

Reply:
if moved to the start then we may miss handling of node Add events:
- as k8s.Probe Start() first starts the links
- then k8s.Probe.Start() Starts the subprobes in *random* order (looping via a go map works that way).

Thus if for example `namespace` will start before `cluster` then clusterLinker.GetBALinks() would get a `namespace` creation event *before* the singleton cluster node exists and thus would not be able to add the edge. As we are not implementing `clusterLinker.GetABLinks()` we have no chance to add the missing edge ....